File: go/networking/whois_search.go
// Whois Search Tool — query and parse domain registration data.
package networking

import (
	"fmt"
	"strings"
)

// WhoisContact holds contact info from a whois record.
type WhoisContact struct {
	Name         string
	Organization string
	Email        string
	City         string
	State        string
	Country      string
}

// Summary returns a one-line contact summary.
func (c WhoisContact) Summary() string {
	var parts []string
	for _, s := range []string{c.Name, c.Organization, c.City, c.Country} {
		if s != "" {
			parts = append(parts, s)
		}
	}
	if len(parts) == 0 {
		return "(redacted)"
	}
	return strings.Join(parts, ", ")
}

// WhoisRecord holds parsed whois data.
type WhoisRecord struct {
	Domain      string
	Registrar   string
	Registrant  *WhoisContact
	NameServers []string
	Status      []string
	CreatedDate string
	UpdatedDate string
	ExpiryDate  string
	DNSSEC      string
	RawText     string
}

// TLD returns the top-level domain.
func (r WhoisRecord) TLD() string {
	parts := strings.Split(r.Domain, ".")
	if len(parts) > 1 {
		return parts[len(parts)-1]
	}
	return ""
}

// Summary returns a multi-line summary.
func (r WhoisRecord) WhoIsSummary() string {
	var b strings.Builder
	fmt.Fprintf(&b, "Domain: %s\n", r.Domain)
	if r.Registrar != "" {
		fmt.Fprintf(&b, "  Registrar: %s\n", r.Registrar)
	}
	if r.Registrant != nil {
		fmt.Fprintf(&b, "  Registrant: %s\n", r.Registrant.Summary())
	}
	if len(r.NameServers) > 0 {
		fmt.Fprintf(&b, "  Name Servers: %s\n", strings.Join(r.NameServers, ", "))
	}
	if r.CreatedDate != "" {
		fmt.Fprintf(&b, "  Created: %s\n", r.CreatedDate)
	}
	if r.ExpiryDate != "" {
		fmt.Fprintf(&b, "  Expires: %s\n", r.ExpiryDate)
	}
	if len(r.Status) > 0 {
		fmt.Fprintf(&b, "  Status: %s\n", strings.Join(r.Status, ", "))
	}
	return b.String()
}

func findWhoisField(raw, key string) string {
	keyLower := strings.ToLower(key)
	for _, line := range strings.Split(raw, "\n") {
		if strings.HasPrefix(strings.ToLower(strings.TrimSpace(line)), keyLower) {
			if idx := strings.Index(line, ":"); idx >= 0 {
				return strings.TrimSpace(line[idx+1:])
			}
		}
	}
	return ""
}

func findAllWhoisFields(raw, key string) []string {
	var results []string
	keyLower := strings.ToLower(key)
	for _, line := range strings.Split(raw, "\n") {
		if strings.HasPrefix(strings.ToLower(strings.TrimSpace(line)), keyLower) {
			if idx := strings.Index(line, ":"); idx >= 0 {
				v := strings.TrimSpace(line[idx+1:])
				if v != "" {
					results = append(results, v)
				}
			}
		}
	}
	return results
}

// ParseWhoisText parses raw whois text into a WhoisRecord.
func ParseWhoisText(domain, raw string) WhoisRecord {
	r := WhoisRecord{
		Domain:  strings.ToLower(domain),
		RawText: raw,
	}
	r.Registrar = findWhoisField(raw, "Registrar")
	r.DNSSEC = findWhoisField(raw, "DNSSEC")
	r.CreatedDate = findWhoisField(raw, "Creation Date")
	if r.CreatedDate == "" {
		r.CreatedDate = findWhoisField(raw, "Created Date")
	}
	r.UpdatedDate = findWhoisField(raw, "Updated Date")
	r.ExpiryDate = findWhoisField(raw, "Registry Expiry Date")
	if r.ExpiryDate == "" {
		r.ExpiryDate = findWhoisField(raw, "Expiration Date")
	}
	for _, ns := range findAllWhoisFields(raw, "Name Server") {
		r.NameServers = append(r.NameServers, strings.ToLower(ns))
	}
	r.Status = findAllWhoisFields(raw, "Domain Status")
	if len(r.Status) == 0 {
		r.Status = findAllWhoisFields(raw, "Status")
	}
	reg := WhoisContact{
		Name:         findWhoisField(raw, "Registrant Name"),
		Organization: findWhoisField(raw, "Registrant Organization"),
		Email:        findWhoisField(raw, "Registrant Email"),
		Country:      findWhoisField(raw, "Registrant Country"),
		City:         findWhoisField(raw, "Registrant City"),
		State:        findWhoisField(raw, "Registrant State"),
	}
	if reg.Name != "" || reg.Organization != "" || reg.Email != "" {
		r.Registrant = &reg
	}
	return r
}

// WhoisSearchClient queries and caches whois data.
type WhoisSearchClient struct {
	cache     map[string]WhoisRecord
	simulated map[string]string
}

// NewWhoisSearchClient creates a new client.
func NewWhoisSearchClient() *WhoisSearchClient {
	return &WhoisSearchClient{
		cache:     make(map[string]WhoisRecord),
		simulated: make(map[string]string),
	}
}

// AddSimulated adds a mock whois response.
func (c *WhoisSearchClient) AddSimulated(domain, raw string) {
	c.simulated[strings.ToLower(domain)] = raw
}

// Query fetches whois data for a domain.
func (c *WhoisSearchClient) Query(domain string) WhoisRecord {
	domain = strings.ToLower(domain)
	if r, ok := c.cache[domain]; ok {
		return r
	}
	raw := c.simulated[domain]
	r := ParseWhoisText(domain, raw)
	c.cache[domain] = r
	return r
}

// RegistrarBreakdown counts domains per registrar.
func (c *WhoisSearchClient) RegistrarBreakdown(domains []string) map[string]int {
	counts := make(map[string]int)
	for _, d := range domains {
		r := c.Query(d)
		key := r.Registrar
		if key == "" {
			key = "unknown"
		}
		counts[key]++
	}
	return counts
}

// CompareWhoisRecords returns changed fields between two records.
func CompareWhoisRecords(before, after WhoisRecord) map[string][2]string {
	changes := make(map[string][2]string)
	if before.Registrar != after.Registrar {
		changes["registrar"] = [2]string{before.Registrar, after.Registrar}
	}
	if strings.Join(before.NameServers, ",") != strings.Join(after.NameServers, ",") {
		changes["name_servers"] = [2]string{
			strings.Join(before.NameServers, ","),
			strings.Join(after.NameServers, ","),
		}
	}
	if before.ExpiryDate != after.ExpiryDate {
		changes["expiry_date"] = [2]string{before.ExpiryDate, after.ExpiryDate}
	}
	return changes
}

File: python/networking/whois_search.py
"""Whois Search Tool — query and parse domain registration data."""

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional
import re


@dataclass
class WhoisContact:
    """Contact information from a whois record."""
    name: str = ""
    organization: str = ""
    email: str = ""
    phone: str = ""
    street: str = ""
    city: str = ""
    state: str = ""
    postal_code: str = ""
    country: str = ""

    def summary(self) -> str:
        parts = [p for p in [self.name, self.organization, self.city, self.country] if p]
        return ", ".join(parts) if parts else "(redacted)"


@dataclass
class WhoisRecord:
    """Parsed whois record for a domain."""
    domain: str
    registrar: str = ""
    registrant: Optional[WhoisContact] = None
    admin_contact: Optional[WhoisContact] = None
    tech_contact: Optional[WhoisContact] = None
    name_servers: list[str] = field(default_factory=list)
    status: list[str] = field(default_factory=list)
    created_date: Optional[datetime] = None
    updated_date: Optional[datetime] = None
    expiry_date: Optional[datetime] = None
    dnssec: str = ""
    raw_text: str = ""

    @property
    def tld(self) -> str:
        parts = self.domain.rsplit(".", 1)
        return parts[-1] if len(parts) > 1 else ""

    @property
    def age_days(self) -> Optional[int]:
        if self.created_date:
            return (datetime.now(timezone.utc) - self.created_date).days
        return None

    @property
    def days_until_expiry(self) -> Optional[int]:
        if self.expiry_date:
            return (self.expiry_date - datetime.now(timezone.utc)).days
        return None

    @property
    def is_expired(self) -> bool:
        if self.expiry_date:
            return self.expiry_date < datetime.now(timezone.utc)
        return False

    def summary(self) -> str:
        lines = [f"Domain: {self.domain}"]
        if self.registrar:
            lines.append(f"  Registrar: {self.registrar}")
        if self.registrant:
            lines.append(f"  Registrant: {self.registrant.summary()}")
        if self.name_servers:
            lines.append(f"  Name Servers: {', '.join(self.name_servers)}")
        if self.created_date:
            lines.append(f"  Created: {self.created_date.strftime('%Y-%m-%d')}")
        if self.expiry_date:
            days = self.days_until_expiry
            label = f"({days} days)" if days and days > 0 else "(EXPIRED)"
            lines.append(f"  Expires: {self.expiry_date.strftime('%Y-%m-%d')} {label}")
        if self.status:
            lines.append(f"  Status: {', '.join(self.status)}")
        return "\n".join(lines)


def parse_whois_text(domain: str, raw: str) -> WhoisRecord:
    """Parse raw whois text into a structured WhoisRecord."""
    record = WhoisRecord(domain=domain, raw_text=raw)

    def find_field(pattern: str) -> str:
        m = re.search(pattern, raw, re.IGNORECASE | re.MULTILINE)
        return m.group(1).strip() if m else ""

    def find_date(pattern: str) -> Optional[datetime]:
        val = find_field(pattern)
        if not val:
            return None
        for fmt in ("%Y-%m-%dT%H:%M:%SZ", "%Y-%m-%d", "%d-%b-%Y", "%Y-%m-%dT%H:%M:%S%z"):
            try:
                dt = datetime.strptime(val, fmt)
                if dt.tzinfo is None:
                    dt = dt.replace(tzinfo=timezone.utc)
                return dt
            except ValueError:
                continue
        return None

    record.registrar = find_field(r"Registrar:\s*(.+)")
    record.dnssec = find_field(r"DNSSEC:\s*(.+)")

    # Name servers
    ns_matches = re.findall(r"Name Server:\s*(.+)", raw, re.IGNORECASE)
    record.name_servers = [ns.strip().lower() for ns in ns_matches if ns.strip()]

    # Status
    status_matches = re.findall(r"(?:Domain )?Status:\s*(.+)", raw, re.IGNORECASE)
    record.status = [s.strip() for s in status_matches if s.strip()]

    # Dates
    record.created_date = find_date(r"Creat(?:ion|ed) Date:\s*(.+)")
    record.updated_date = find_date(r"Updated Date:\s*(.+)")
    record.expiry_date = find_date(r"(?:Registry Expiry|Expir(?:ation|y)) Date:\s*(.+)")

    # Registrant contact
    registrant = WhoisContact()
    registrant.name = find_field(r"Registrant Name:\s*(.+)")
    registrant.organization = find_field(r"Registrant Organization:\s*(.+)")
    registrant.email = find_field(r"Registrant Email:\s*(.+)")
    registrant.country = find_field(r"Registrant Country:\s*(.+)")
    registrant.city = find_field(r"Registrant City:\s*(.+)")
    registrant.state = find_field(r"Registrant State/Province:\s*(.+)")
    if any([registrant.name, registrant.organization, registrant.email]):
        record.registrant = registrant

    return record


class WhoisClient:
    """Whois client that queries and parses domain registration data.

    The actual whois query requires a TCP connection to port 43 of the
    appropriate whois server. This module provides the parsing layer and
    a simulated database for in-process use.
    """

    def __init__(self) -> None:
        self._cache: dict[str, WhoisRecord] = {}
        self._simulated_db: dict[str, str] = {}

    def add_simulated(self, domain: str, raw_text: str) -> None:
        """Add a simulated whois response for testing."""
        self._simulated_db[domain.lower()] = raw_text

    def query(self, domain: str) -> WhoisRecord:
        """Query whois data for a domain."""
        domain = domain.lower().strip()
        if domain in self._cache:
            return self._cache[domain]
        raw = self._simulated_db.get(domain, "")
        record = parse_whois_text(domain, raw)
        self._cache[domain] = record
        return record

    def bulk_query(self, domains: list[str]) -> list[WhoisRecord]:
        """Query multiple domains."""
        return [self.query(d) for d in domains]

    def check_expiry(self, domains: list[str], warn_days: int = 30) -> list[WhoisRecord]:
        """Return domains expiring within warn_days."""
        results = []
        for d in domains:
            r = self.query(d)
            if r.days_until_expiry is not None and 0 < r.days_until_expiry <= warn_days:
                results.append(r)
        return results

    def registrar_breakdown(self, domains: list[str]) -> dict[str, int]:
        """Count domains per registrar."""
        counts: dict[str, int] = {}
        for d in domains:
            r = self.query(d)
            key = r.registrar or "unknown"
            counts[key] = counts.get(key, 0) + 1
        return counts


def compare_records(before: WhoisRecord, after: WhoisRecord) -> dict[str, tuple]:
    """Compare two whois records for the same domain, returning changed fields."""
    changes: dict[str, tuple] = {}
    if before.registrar != after.registrar:
        changes["registrar"] = (before.registrar, after.registrar)
    if before.name_servers != after.name_servers:
        changes["name_servers"] = (before.name_servers, after.name_servers)
    if before.expiry_date != after.expiry_date:
        changes["expiry_date"] = (before.expiry_date, after.expiry_date)
    if before.status != after.status:
        changes["status"] = (before.status, after.status)
    return changes


# --- demo ---
if __name__ == "__main__":
    client = WhoisClient()

    # Simulate a whois response
    client.add_simulated("eckenrodemuziekopname.com", """
Domain Name: eckenrodemuziekopname.com
Registrar: Namecheap, Inc.
Creation Date: 2023-06-15T00:00:00Z
Updated Date: 2025-06-15T00:00:00Z
Registry Expiry Date: 2027-06-15T00:00:00Z
Registrant Name: Nathan Eckenrode
Registrant Organization: Eckenrode Muziekopname
Registrant Country: US
Registrant City: Jacksonville
Registrant State/Province: Florida
Name Server: ns1.digitalocean.com
Name Server: ns2.digitalocean.com
Name Server: ns3.digitalocean.com
Domain Status: clientTransferProhibited
DNSSEC: unsigned
""")

    record = client.query("eckenrodemuziekopname.com")
    print(record.summary())
